runbot_buildout/models/runbot_build.py

```diff
@@ -0,0 +1,335 @@
+                    dirname = os.path.dirname(module)
+                    basename = os.path.basename(dirname)
+                    if os.path.exists(build._server('addons', basename)):
+                        build._log(
+                            'Building environment',
+                            'Duplicate module "%s"' % basename
+                        )
+                        shutil.rmtree(
+                            build.server('addons', basename),
+                            ignore_errors=True
+                        )
+                    shutil.move(dirname, build._server('addons'))
+                    available_modules.append(basename)
+            modules_to_test = (
+                (build.branch_id.modules or '') + ',' +
+                (build.repo_id.modules or '')
+            )
+            modules_to_test = filter(None, modules_to_test.split(','))
+            explicit_modules = set(modules_to_test)
+            if build.repo_id.modules_auto == 'all':
+                modules_to_test += available_modules
+
+            # TODO: write this in _job_init if repo.modules_auto == 'repo'
+            build.write({
+                'modules': ','.join(self._filter_modules(
+                    modules_to_test, set(available_modules), explicit_modules,
+                ))
+            })
+        return super(RunbotBuild, self)._job_10_test_base(
+            build, lock_path, log_path
+        )
+
+    @api.model
+    def _job_20_test_all(self, build, lock_path, log_path):
+        if build.repo_id.uses_buildout and build.branch_id.buildout_version:
+            build._log('buildout', 'Running buildout')
+            return build._spawn_buildout(
+                [build._path('bin/buildout'), '-N', '-q'],
+                lock_path, log_path
+            )
+        return super(RunbotBuild, self)._job_20_test_all(
+            build, lock_path, log_path
+        )
+
+    @api.model
+    def _job_30_run(self, build, lock_path, log_path):
+        if build.repo_id.uses_buildout and build.branch_id.buildout_version:
+            if self._check_buildout_success(build):
+                build.write({'result': 'ok'})
+            return -2
+        return super(RunbotBuild, self)._job_30_run(
+            build, lock_path, log_path
+        )
+
+    @api.model
+    def _check_buildout_success(self, build):
+        result = None
+        if os.path.exists(
+            build._path('bin/start_%s') % build.repo_id.buildout_section
+        ):
+            build._log('buildout', 'Buildout succeeded')
+            result = True
+        else:
+            build._log('buildout', 'Buildout failed')
+            build.write({'result': 'ko'})
+            result = False
+        if not result:
+            build._github_status()
+        return result
+
+    @api.multi
+    def _spawn_buildout(self, cmd, lock_path, log_path):
+        self.ensure_one()
+        out = open(log_path, "w")
```

Comment:
When is this file closed?

Reply:
by now a few lines below, thanks for this one

Reply:
Ok, What about use `with open(log_path, "w")` in order to ensure that this file is closed if there is an error in the middle.